Add footers with author names, create date and first commit hash ID

rebeccabilbro opened the pull request:
This PR adds headers with author names, create date and first commit hash ID. The script scans a local directory for git repository metadata, and if files are found that have headers with and ID line stub, adds the correct initial commit hash to the ID line. Files in `vendor` and those without the .go file extension are ignored. Moving forward, this will require us to add the following header to all new Go files:

```go
/*
Package example is an example package.

Author:  Ykceb Orblib <ykceb@kansaslabs.com>
Created: Sat Sep 14 13:26:53 EDT 2019

Copyright (C) 2019 Kansas Labs
For license information, see LICENSE.txt

ID: example.go [] ykceb@orblib.com $ 
*/
```

V1 of the ID script for Kansas Golang packages can be found in [this gist](https://gist.github.com/rebeccabilbro/ca826de48434398921dfd951eeeeb770).

rebeccabilbro committed:
adds headers to baleen files with authors, license information and commit ID

rebeccabilbro committed:
author update

bbengfort commented:
This looks really great and the ID script is also very good. Unfortunately, I've just discovered that this may not work as we'd hoped. Since go treats all .go files as a single package as opposed to independent modules (like in Python) the docstrings are all used to generate the documentation and are appended together in Godoc as follows:

![Screenshot 2019-09-26 17 41 02](https://user-images.githubusercontent.com/745966/65727232-1fc83f00-e085-11e9-9d13-1581cd247057.png)

I'm super bummed about this, and not sure if there is a fix (e.g. using `//` comments instead of `/* */`? We may have to figure out a convention where we remove the `ID` lines and have only one of these author/copyright docstrings per-directory (e.g. in the `fetch.go` or `baleen.go` files). It is a bit confusing because some licenses like the BSD license require the license to be put in every source file (e.g. why I like that see LICENSE.txt line in the header). I looked for some examples on godoc but couldn't find any. :-(

Any thoughts?

rebeccabilbro committed:
move ids and copyright info to footers

rebeccabilbro commented:
I've modified the PR to move copyright information to footer docstrings, which go doc ignores entirely. I realize this is not ideal as we are used to having and looking for such information at the top of our Python files, but unfortunately go doc doesn't do deduplication of the headers across packages, so for now this may be the best way to ensure that copyright and license information appears on every file. 

Moving forward, this would require us to:

1. Add a unique header to each package file:
```go
/*
Package example is an example package.
*/
```
_Note: These should be unique within a package, since go doc appears only to automatically deduplicate the headers for top-level package files._

2. Add the following footer template:
```go
/*
Author:  Ykceb Orblib <ykceb@kansaslabs.com>
Created: Sat Sep 14 13:26:53 EDT 2019

Copyright (C) 2019 Kansas Labs
For license information, see LICENSE.txt

ID: example.go [] ykceb@orblib.com $ 
*/
```

Updated ID script [here](https://gist.github.com/rebeccabilbro/ca826de48434398921dfd951eeeeb770).

bbengfort reviewed (approved):
This is pretty ingenious - I didn't think of code footers, but this definitely solves the problem! I think it's going to take a bit for me to get my head around it, but what a clever solution; nice one!

rebeccabilbro committed:
Merge branch 'develop' of github.com:kansaslabs/baleen into add-headers

rebeccabilbro committed:
adds footer to opml utils package
